src/agora/agora.cc

```diff
@@ -407,21 +407,34 @@ void Agora::Start() {
               this->demul_counters_.CompleteTask(frame_id, symbol_id);
 
           if (last_demul_task == true) {
-            ScheduleCodeblocks(EventType::kDecode, frame_id, symbol_id);
+            if (kUplinkHardDemod == false) {
+              ScheduleCodeblocks(EventType::kDecode, Direction::kUplink,
+                                 frame_id, symbol_id);
+            }
             PrintPerSymbolDone(PrintType::kDemul, frame_id, symbol_id);
             bool last_demul_symbol =
                 this->demul_counters_.CompleteSymbol(frame_id);
             if (last_demul_symbol == true) {
-              this->demul_counters_.Reset(frame_id);
               max_equaled_frame_ = frame_id;
-              if (cfg->BigstationMode() == false) {
-                assert(cur_sche_frame_id_ == frame_id);
+              this->stats_->MasterSetTsc(TsType::kDemulDone, frame_id);
+              PrintPerFrameDone(PrintType::kDemul, frame_id);
+              if (kUplinkHardDemod == true) {
```

Comment:
We need to add some comments here about what the difference between hard demodulation and soft demodulation is.